timelock/src/lib.rs

```diff
@@ -0,0 +1,212 @@
+            BigInt::zero(&env),
+            Symbol::from_str("deposit"),
+            (
+                &from_id,
+                &token,
+                &amount,
+                &claimants,
+                &time_bound_kind,
+                timestamp,
+            )
+                .into_val(&env),
+        );
+        // Transfer token to this contract address.
+        transfer_from(&env, &token, &from_id, &get_contract_id(&env), &amount);
+        // Store all the necessary balance to allow one of the claimants to claim it.
+        env.contract_data().set(
+            DataKey::Balance,
+            ClaimableBalance {
+                token,
+                amount,
+                time_bound: TimeBound {
+                    kind: time_bound_kind,
+                    timestamp,
+                },
+                claimants,
+            },
+        );
+    }
+
+    pub fn claim(env: Env, claimant: Signature) {
+        if !is_initialized(&env) {
+            panic!("contract hasn't been initialized");
+        }
+        let claimable_balance: ClaimableBalance =
+            env.contract_data().get_unchecked(DataKey::Balance).unwrap();
+
+        if !check_time_bound(&env, &claimable_balance.time_bound) {
+            panic!("time predicate is not fulfilled");
+        }
+
+        let claimant_id = claimant.get_identifier(&env);
+        let claimants = &claimable_balance.claimants;
+        // Check if claimaint is one of the allowed claimants.
+        // `contains` is not implemented for Soroban vec (yet).
+        if !claimants.iter().any(|c| c.unwrap() == claimant_id) {
+            panic!("claimant is not allowed to claim this balance");
+        }
+        // Authenticate claimant with nonce of zero, so that this may be
+        // successfully called just once.
+        // For simplicity, depositor can't be the claimant.
+        check_auth(
+            &env,
+            &NonceForSignature(claimant),
+            BigInt::zero(&env),
+            Symbol::from_str("claim"),
+            (&claimant_id,).into_val(&env),
+        );
+        // Transfer the stored amount of token to claimant after passing
+        // all the checks.
+        transfer_to(
+            &env,
+            &claimable_balance.token,
+            &claimant_id,
+            &claimable_balance.amount,
+        );
+        // Cleanup unnecessary balance entry.
+        env.contract_data().remove(DataKey::Balance);
+    }
+}
+
+fn is_initialized(env: &Env) -> bool {
+    env.contract_data().has(DataKey::Balance)
+}
+
+fn get_contract_id(e: &Env) -> Identifier {
+    Identifier::Contract(e.get_current_contract().into())
+}
+
+fn transfer_from(
+    e: &Env,
+    token_id: &BytesN<32>,
+    from: &Identifier,
+    to: &Identifier,
+    amount: &BigInt,
+) {
+    let client = TokenClient::new(&e, token_id);
+    let nonce = client.nonce(&get_contract_id(&e));
+    client.xfer_from(&Signature::Contract, &nonce, &from, &to, &amount);
```

Comment:
The nonce for the Contract Signature type is expected to be 0, so you can just pass that in instead of calling the nonce function.

Reply:
I just realized the `single_offer_xfer_from` does the same thing. I'll make sure to fix that.

Reply:
Both calls use contract auth, so only the first one can be 0. I guess we could use 0 for this function, but I'm not sure it's reasonable for a helper function that doesn't necessarily know when is it going to be called.

Reply:
The auth sdk expects the passed in nonce to be [zero](https://github.com/stellar/rs-soroban-sdk/blob/6d129bf6a9250d54d67ee851ea027d4a4f892673/soroban-auth/src/lib.rs#L128) for `Signature::Contract` because it isn't used, so there's no reason to do this cross contract call. 

The way the nonce is passed into `check_auth` isn't ideal because of this. Maybe our solution to https://github.com/stellar/rs-soroban-sdk/issues/533 will improve this as well.

Reply:
I see, that's a bit confusing indeed. Changed nonce to zero.